Automattic/woocommerce-payments
# Replace direct DB query to posts and postmeta tables by the corresponding WC APIs

Fixes #4151 

#### Changes proposed in this Pull Request

To retrieve an order using a custom metadata, we have to add a filter, as recommended here: https://github.com/woocommerce/woocommerce/wiki/wc_get_orders-and-WC_Order_Query#adding-custom-parameter-support

[orders_with_charge_id_from_charge_ids()](https://github.com/Automattic/woocommerce-payments/blob/ca0aa7a9160647e3612bffc8b9e0f935bea3acd1/includes/class-wc-payments-db.php#L46-L52) - Retrieve orders from the DB using a list of corresponding Stripe charge IDs.

- Refactored to use `wc_get_orders` function and a custom query variable filter to use the `_charge_id` metadata key/value
- Now only returns a list of orders instead of a list of [ order + charge_id ] since the _charge_id metadata can be retrieved directly from the order

 [order_id_from_meta_key_value()](https://github.com/Automattic/woocommerce-payments/blob/ca0aa7a9160647e3612bffc8b9e0f935bea3acd1/includes/class-wc-payments-db.php#L93-L99) - Retrieve the latest order ID from the DB using a meta key value pair.

- Refactored to use `wc_get_orders` function and a custom query variable filter to use the metadata key/value provided

#### Testing instructions

On the current WooCommerce version:

- If you navigate to Payments > Transactions, you should see the `Order #` column populated as usual (order number with a link pointing to the order), as well as the `Customer` column
- If you navigate to Payments > Disputes, you should see the `Order #` column populated as usual (order number with a link pointing to the order)
- Make an order using [a special CC for disputes](https://stripe.com/docs/testing#disputes)
  - the [`payment_intent.succeeded`](https://github.com/Automattic/woocommerce-payments/blob/df22af0e86e9e99912e7902419685a67f32999c3/includes/class-wc-payments-webhook-processing-service.php#L365) webhook event should be processed as usual (order saved with the right metadata, order note for payment success added, the status is set to payment_complete, ...)
   - the [`charge.dispute.created`](https://github.com/Automattic/woocommerce-payments/blob/df22af0e86e9e99912e7902419685a67f32999c3/includes/class-wc-payments-webhook-processing-service.php#L423) webhook event should be processed as usual (dispute created note is added to the order, the status is set to on-hold, ...)

    - Same check to be made for the `payment_intent.payment_failed` webhook event (payment failure note is added to the order, the status is set to failed, ...), if you use a wrong CC info

That's a sample of the tests to check. There are other webhooks involving the `order_from_charge_id` function that was refactored (dispute updated, dispute closed, refund updated, expired uncaptured payment). This function is also involved in displaying a dispute or list of disputes, a transaction or list of transactions, a charge or updating or closing a dispute.

The same checks should be made on the different test sites for COT testing (COT disabled, COT enabled without sync, COT enabled with sync). It should work in all environments.
More info [here](https://github.com/Automattic/woocommerce/wiki/COT-Upgrade-Recipe-Book#easy-way) for COT testing.

-------------------


**Post merge**
